[zh]sync concept of statefulset (#33364)
kubernetes/website: content/zh/docs/concepts/workloads/controllers/statefulset.md

StatefulSet 的命名需要遵循[DNS 子域名](/zh/docs/concepts/overview/working-with-objects/names#dns-subdomain-names)规范。

<!--
### Pod Selector
-->
### Pod 选择算符     {#pod-selector}

<!--
You must set the `.spec.selector` field of a StatefulSet to match the labels of its `.spec.template.metadata.labels`. Failing to specify a matching Pod Selector will result in a validation error during StatefulSet creation.
-->
你必须设置 StatefulSet 的 `.spec.selector` 字段，使之匹配其在
`.spec.template.metadata.labels` 中设置的标签。
未指定匹配的 Pod 选择器将在创建 StatefulSet 期间导致验证错误。

<!-- 
### Volume Claim Templates

You can set the  `.spec.volumeClaimTemplates` which can provide stable storage using [PersistentVolumes](/docs/concepts/storage/persistent-volumes/) provisioned by a PersistentVolume Provisioner.
-->
### 卷申领模版  {#volume-claim-templates}

你可以设置 `.spec.volumeClaimTemplates`，

> 它可以使用 PersistentVolume 供应商供应的

Reviewer: provisioner 和 provider 还不太一样，provision 作动词的时候是 make something ready to use 的意思，和“供应”也不太相同。

```suggestion
它可以使用 PersistentVolume 制备程序所准备的
```

[PersistentVolumes](/zh/docs/concepts/storage/persistent-volumes/) 来提供稳定的存储。

<!-- ### Minimum ready seconds -->
### 最短就绪秒数 {#minimum-ready-seconds}

{{< feature-state for_k8s_version="v1.23" state="beta" >}}

<!-- 
`.spec.minReadySeconds` is an optional field that specifies the minimum number of seconds for which a newly
created Pod should be ready without any of its containers crashing, for it to be considered available.
Please note that this feature is beta and enabled by default. Please opt out by unsetting the StatefulSetMinReadySeconds flag, if you don't
want this feature to be enabled. This field defaults to 0 (the Pod will be considered 
available as soon as it is ready). To learn more about when a Pod is considered ready, see [Container Probes](/docs/concepts/workloads/pods/pod-lifecycle/#container-probes).
-->
`.spec.minReadySeconds` 是一个可选字段，
它指定新创建的 Pod 应该准备好且其任何容器不崩溃的最小秒数，以使其被视为可用。
请注意，此功能是测试版，默认启用。如果你不希望启用此功能，
请通过取消设置 StatefulSetMinReadySeconds 标志来选择退出。
该字段默认为 0（Pod 准备就绪后将被视为可用）。

> 要了解有关何时认为 Pod 准备就绪的更多信息，请参阅
> [容器探针](/zh/docs/concepts/workloads/pods/pod-lifecycle/#container-probes)。

Reviewer: ```suggestion
要了解有关何时认为 Pod 准备就绪的更多信息，
请参阅[容器探针](/zh/docs/concepts/workloads/pods/pod-lifecycle/#container-probes)。
```

恢复模板后，还必须删除 StatefulSet 尝试使用错误的配置来运行的 Pod。这样，
StatefulSet 才会开始使用被还原的模板来重新创建 Pod。

<!-- ## PersistentVolumeClaim retention -->

> ## PersistentVolumeClaim 保持  {#persistentvolumeclaim-retention}

Reviewer: retention 译为“保留”是否更好？

Author: 赞同

StatefulSet:
-->
在 StatefulSet 的生命周期中，可选字段 
`.spec.persistentVolumeClaimRetentionPolicy` 控制是否删除以及如何删除 PVC。
使用该字段，你必须启用 `StatefulSetAutoDeletePVC`
[特性门控](/zh/docs/reference/command-line-tools-reference/feature-gates/)。
启用后，你可以为每个 StatefulSet 配置两个策略：

<!-- 
`whenDeleted`
: configures the volume retention behavior that applies when the StatefulSet is deleted

`whenScaled`
: configures the volume retention behavior that applies when the replica count of
  the StatefulSet   is reduced; for example, when scaling down the set.
  
For each policy that you can configure, you can set the value to either `Delete` or `Retain`.
-->
`whenDeleted`
: 配置删除 StatefulSet 时应用的卷保留行为

`whenScaled`
: 配置当 StatefulSet 的副本数减少时应用的卷保留行为；例如，缩小集合时。
  
对于你可以配置的每个策略，你可以将值设置为 `Delete` 或 `Retain`。

<!-- 
`Delete`
: The PVCs created from the StatefulSet `volumeClaimTemplate` are deleted for each Pod
  affected by the policy. With the `whenDeleted` policy all PVCs from the
  `volumeClaimTemplate` are deleted after their Pods have been deleted. With the
  `whenScaled` policy, only PVCs corresponding to Pod replicas being scaled down are
  deleted, after their Pods have been deleted.
-->
`Delete`

> : 对于受策略影响的每个 Pod，从 StatefulSet 的 `volumeClaimTemplate` 字段创建的 PVC 都会被删除。

Reviewer: ```suggestion
: 对于受策略影响的每个 Pod，基于 StatefulSet 的 `volumeClaimTemplate` 字段创建的 PVC 都会被删除。
```

The default for policies is `Retain`, matching the StatefulSet behavior before this new feature.

Here is an example policy.
-->
请记住，这些策略**仅**适用于由于 StatefulSet 被删除或被缩小而被删除的 Pod。
例如，如果与 StatefulSet 关联的 Pod 由于节点故障而失败，
并且控制平面创建了替换 Pod，则 StatefulSet 保留现有的 PVC。
现有卷不受影响，集群会将其附加到新 Pod 即将启动的节点上。

策略的默认值为 `Retain`，与此新功能之前的 StatefulSet 行为相匹配。

这是一个示例策略。

```yaml
apiVersion: apps/v1
kind: StatefulSet
...
spec:
  persistentVolumeClaimRetentionPolicy:
    whenDeleted: Retain
    whenScaled: Delete
...
```

<!--
The StatefulSet {{<glossary_tooltip text="controller" term_id="controller">}} adds [owner
references](/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications)
to its PVCs, which are then deleted by the {{<glossary_tooltip text="garbage collector"
term_id="garbage-collection">}} after the Pod is terminated. This enables the Pod to
cleanly unmount all volumes before the PVCs are deleted (and before the backing PV and
volume are deleted, depending on the retain policy).  When you set the `whenDeleted`
policy to `Delete`, an owner reference to the StatefulSet instance is placed on all PVCs
associated with that StatefulSet. 
-->

> StatefulSet {{<glossary_tooltip text="控制器" term_id="controller">}}添加了
> [属主引用](/zh/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications)
> 到它的 PVC，然后在 Pod 终止后被{{<glossary_tooltip text="垃圾回收器" term_id="garbage-collection">}}删除。

Reviewer: ```suggestion
StatefulSet {{<glossary_tooltip text="控制器" term_id="controller">}}为其 PVC 添加了
[属主引用](/zh/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications)，
这些 PVC 在 Pod 终止后被{{<glossary_tooltip text="垃圾回收器" term_id="garbage-collection">}}删除。
```

这使 Pod 能够在删除 PVC 之前（以及在删除后备 PV 和卷之前，取决于保留策略）干净地卸载所有卷。
当你设置 `whenDeleted` 删除策略，对 StatefulSet 实例的属主引用放置在与该 StatefulSet 关联的所有 PVC 上。

<!--
The `whenScaled` policy must delete PVCs only when a Pod is scaled down, and not when a
Pod is deleted for another reason. When reconciling, the StatefulSet controller compares
its desired replica count to the actual Pods present on the cluster. Any StatefulSet Pod
whose id greater than the replica count is condemned and marked for deletion. If the
`whenScaled` policy is `Delete`, the condemned Pods are first set as owners to the
associated StatefulSet template PVCs, before the Pod is deleted. This causes the PVCs
to be garbage collected after only the condemned Pods have terminated.
-->
`whenScaled` 策略必须仅在 Pod 缩减时删除 PVC，而不是在 Pod 因其他原因被删除时删除。

> 协调时，StatefulSet 控制器将其所需的副本数与集群上实际存在的 Pod 进行比较。

Reviewer: ```suggestion
执行协调操作时，StatefulSet 控制器将其所需的副本数与集群上实际存在的 Pod 进行比较。
```

> 任何 StatefulSet 的 Pod，其 id 大于副本数将被废弃并标记为删除。

Reviewer: ```suggestion
对于 StatefulSet 中的所有 Pod 而言，如果其 ID 大于副本数，则将被废弃并标记为需要删除。
```

如果 `whenScaled` 策略是 `Delete`，则在删除 Pod 之前，

> 首先将已销毁的 Pod 设置为关联的 StatefulSet 模板 PVC 的所有者。

Reviewer: ```suggestion
首先将已销毁的 Pod 设置为与 StatefulSet 模板 对应的 PVC 的属主。
```

这会导致 PVC 仅在已废弃的 Pod 终止后被垃圾收集。

<!-- 
This means that if the controller crashes and restarts, no Pod will be deleted before its
owner reference has been updated appropriate to the policy. If a condemned Pod is
force-deleted while the controller is down, the owner reference may or may not have been
set up, depending on when the controller crashed. It may take several reconcile loops to
update the owner references, so some condemned Pods may have set up owner references and
other may not. For this reason we recommend waiting for the controller to come back up,
which will verify owner references before terminating Pods. If that is not possible, the
operator should verify the owner references on PVCs to ensure the expected objects are
deleted when Pods are force-deleted.
-->
这意味着如果控制器崩溃并重新启动，在其属主引用更新到适合策略的 Pod 之前，不会删除任何 Pod。

> 如果在控制器关闭时强制删除了已废弃的 Pod，则属主引用可能已设置，也可能未设置，具体取决于控制器何时崩溃。

Reviewer: ```suggestion
如果在控制器关闭时强制删除了已废弃的 Pod，则属主引用可能已被设置，也可能未被设置，具体取决于控制器何时崩溃。
```

> 更新属主引用可能需要几个协调循环，因此一些已废弃的 Pod 可能已经设置了属主引用，而其他可能没有。

Reviewer: ```suggestion
更新属主引用可能需要几个协调循环，因此一些已废弃的 Pod 可能已经被设置了属主引用，而其他可能没有。
```

> 出于这个原因，我们建议等待控制器恢复，这将在终止 Pod 之前验证属主引用。

Reviewer: ```suggestion
出于这个原因，我们建议等待控制器恢复，控制器将在终止 Pod 之前验证属主引用。
```

如果这不可行，则操作员应验证 PVC 上的属主引用，以确保在强制删除 Pod 时删除预期的对象。

<!-- 
### Replicas

`.spec.replicas` is an optional field that specifies the number of desired Pods. It defaults to 1.

Should you manually scale a deployment, example via `kubectl scale
statefulset statefulset --replicas=X`, and then you update that StatefulSet
based on a manifest (for example: by running `kubectl apply -f
statefulset.yaml`), then applying that manifest overwrites the manual scaling
that you previously did.
-->

> 

Reviewer: 删除 701 行

### 副本数 {#replicas}

> `.spec.replicas` 是一个可选字段，用于指定所需 Pod 的数量。它默认值为 1。

Reviewer: ```suggestion
`.spec.replicas` 是一个可选字段，用于指定所需 Pod 的数量。它的默认值为 1。
```

> 如果你手动扩展部署，例如通过 `kubectl scale statefulset statefulset --replicas=X`，

Reviewer: ```suggestion
如果你手动扩缩已部署的负载，例如通过 `kubectl scale statefulset statefulset --replicas=X`，
```

然后根据清单更新 StatefulSet（例如：通过运行 `kubectl apply -f statefulset.yaml`），

> 然后应用该清单会覆盖你之前所做的手动缩放。

Reviewer: ```suggestion
那么应用该清单的操作会覆盖你之前所做的手动缩放。
```
